solidity/contracts/l2/L2WormholeGateway.sol

```diff
@@ -0,0 +1,257 @@
+        // we check the balance before the ITokenBridge call and the balance
+        // after ITokenBridge call. This way, we are sure this contract received
+        // Wormhole tBTC token in the given amount. This is transparent to
+        // all potential upgrades of ITokenBridge implementation and no other
+        // validations are needed.
+        uint256 balanceBefore = bridgeToken.balanceOf(address(this));
+        bytes memory payload = bridge.completeTransferWithPayload(encodedVm);
+        uint256 balanceAfter = bridgeToken.balanceOf(address(this));
+
+        uint256 amount = balanceAfter - balanceBefore;
+        address receiver = abi.decode(payload, (address));
+        require(receiver != address(0), "0x0 receiver not allowed");
+
+        // We send wormhole tBTC OR mint canonical tBTC. We do not want to send
+        // dust. Sending wormhole tBTC is an exceptional situation and we want
+        // to keep it simple.
+        if (mintedAmount + amount > mintingLimit) {
+            bridgeToken.safeTransfer(receiver, amount);
+        } else {
+            // Function is nonReentrant and we need to update the balance with
+            // a call to completeTransferWithPayload first to know the amount.
+            // slither-disable-next-line reentrancy-benign
+            mintedAmount += amount;
+            tbtc.mint(receiver, amount);
+        }
+
+        // Function is nonReentrant and we need to extract the payload with
+        // a call to completeTransferWithPayload first.
+        // slither-disable-next-line reentrancy-events
+        emit WormholeTbtcReceived(receiver, amount);
+    }
+
+    /// @notice This function is called when the user sends their token from L2.
+    ///         The contract burns the canonical tBTC from the user and sends
+    ///         wormhole tBTC representation over the bridge.
+    ///         Keep in mind that when multiple bridges receive a minting
+    ///         authority on the canonical tBTC, this function may not be able
+    ///         to send all amounts of tBTC through the Wormhole bridge. The
+    ///         capability of Wormhole Bridge to send tBTC from the chain is
+    ///         limited to the amount of tBTC bridged through Wormhole to that
+    ///         chain.
+    /// @dev Requirements:
+    ///      - The sender must have at least `amount` of the canonical tBTC and
+    ///        it has to be approved for L2WormholeGateway.
+    ///      - The L2WormholeGateway must have at least `amount` of the wormhole
+    ///        tBTC.
+    /// @param amount The amount of tBTC to be sent.
+    /// @param recipientChain The Wormhole recipient chain ID.
+    /// @param recipient The address of the recipient in the Wormhole format.
+    /// @param arbiterFee The Wormhole arbiter fee.
+    /// @param nonce The Wormhole nonce used to batch messages together.
+    /// @return The Wormhole sequence number.
+    function sendWormhole(
```

Comment:
Just a question: Suppose the minting limit was exceeded during the `receiveWormhole` call and the user got their wormhole TBTC instead of the canonical one. What if the user wants to go back to L1? The `sendWormhole` cannot be used for that right?

Reply:
Correct, `sendWormhole` can't be used for this but they can go back to L1 working with the Wormhole token bridge contract directly.